cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -10,70 +10,302 @@
+        context.cgmes().substations().forEach(st -> addSubstation(substationAdjacency, st));
+
+        context.cgmes().switches().forEach(sw -> addSwitch(voltageLevelAdjacency, substationAdjacency, sw));
+        context.cgmes().groupedTransformerEnds().forEach((t, tends) -> addEnds(substationAdjacency, tends));
+    }
+
+    private void addVoltageLevel(Map<String, List<String>> voltageLevelAdjacency, PropertyBag vl) {
+        String voltageLevelId = vl.getId(CgmesNames.VOLTAGE_LEVEL);
+        String vId = context.namingStrategy().getId(CgmesNames.VOLTAGE_LEVEL, voltageLevelId);
+        voltageLevelAdjacency.computeIfAbsent(vId, k -> new ArrayList<>());
+    }
+
+    private void addSubstation(Map<String, List<String>> substationAdjacency, PropertyBag sub) {
+        String substationlId = sub.getId(CgmesNames.SUBSTATION);
+        String subId = context.namingStrategy().getId(CgmesNames.SUBSTATION, substationlId);
+        substationAdjacency.computeIfAbsent(subId, k -> new ArrayList<>());
+    }
+
+    // Two different voltageLevels are adjacent if they are connected by a switch
+    // If the corresponding substations are different they are also adjacent
+    private void addSwitch(Map<String, List<String>> voltageLevelAdjacency,
+        Map<String, List<String>> substationAdjacency, PropertyBag sw) {
+
+        CgmesTerminal t1 = context.cgmes().terminal(sw.getId(CgmesNames.TERMINAL + 1));
+        String node1 = context.nodeBreaker() ? t1.connectivityNode() : t1.topologicalNode();
+        String voltageLevelId1 = getVoltageLevelFromNode(node1, t1);
+
+        CgmesTerminal t2 = context.cgmes().terminal(sw.getId(CgmesNames.TERMINAL + 2));
+        String node2 = context.nodeBreaker() ? t2.connectivityNode() : t2.topologicalNode();
+        String voltageLevelId2 = getVoltageLevelFromNode(node2, t2);
+
+        // Null could be received as voltageLevel at the boundary
+        if (voltageLevelId1 == null || voltageLevelId2 == null || voltageLevelId1.equals(voltageLevelId2)) {
+            return;
+        }
+        addAdjacency(voltageLevelAdjacency, voltageLevelId1, voltageLevelId2);
+
+        String substationId1 = context.cgmes().substation(t1, context.nodeBreaker());
+        String substationId2 = context.cgmes().substation(t2, context.nodeBreaker());
+
+        // Null could be received as substation at the boundary
+        if (substationId1 == null || substationId2 == null || substationId1.equals(substationId2)) {
+            return;
+        }
+        addAdjacency(substationAdjacency, substationId1, substationId2);
+    }
+
+    private String getVoltageLevelFromNode(String node, CgmesTerminal t) {
+        String voltageLevelId = null;
+        if (node != null && !context.boundary().containsNode(node)) {
+            voltageLevelId = context.cgmes().voltageLevel(t, context.nodeBreaker());
+        }
+        return voltageLevelId;
+    }
```

Comment:
Maybe this utility function is used elsewhere (maybe it already exists). Could we move it outside this class?

Reply:
Done